File: packages/gg-sandbox-ruby/src/gg/session.rb
# frozen_string_literal: true

module GG
  # End the session.
  #
  # None of these calls stops the program: whatever follows an ending still runs. A program that
  # then fails has its ending revoked along with everything else it decided.
  module Session
    extend Surface::Operations

    # End the session, reporting what was done in a sentence or two.
    #
    # It does not stop the program: whatever follows it still runs. A program that then fails has
    # the ending cancelled and gets another turn.
    #
    # @param summary [String] What was done, in a sentence or two.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:invalid_argument` for a blank summary.
    def self.finish(summary)
      Wire.call("finish", "session", "finish", [summary])
      nil
    end
    operation :finish, "session.finish"

    # Accept the work under review: it meets every completion criterion and stays in scope.
    #
    # This ends the session. It does not stop the program: whatever follows it still runs.
    #
    # @return [nil]
    def self.approve
      Wire.call("approve", "session", "approve", [])
      nil
    end
    operation :approve, "session.approve"

    # Reject the work under review, listing every change that must be made before it can be
    # accepted.
    #
    # This ends the session, and does not stop the program. Each item says what is wrong and what to
    # change.
    #
    # @param items [Array<String>] Every change that must be made before the work can be accepted,
    #   splatted, one per entry: what is wrong, and what to change. It may not be empty.
    # @return [nil]
    # @raise [GG::Core::ApiError] `:invalid_argument` when the list is empty.
    def self.request_changes(*items)
      Wire.call("request_changes", "session", "requestChanges",
                [Check.strings("request_changes", "items", items)])
      nil
    end
    operation :request_changes, "session.request_changes"
  end
end

File: packages/gg-sandbox-rust/src/session.rs
//! End the session, with the ending that belongs to this session's role.
//!
//! A run binds only the group its role has: work is ended by reporting what was done, and a review
//! is ended with a verdict.
//!
//! None of these calls stops the program. Whatever follows an ending still runs, and a program that
//! then fails has its ending revoked along with everything else it decided.

use crate::bindings::test_cabinet::gg::session;
use crate::core::ApiError;
use crate::wire;

/// End the session, reporting what was done in a sentence or two.
///
/// It does not stop the program: whatever follows it still runs. A program that then fails has the
/// ending cancelled and gets another turn.
///
/// # Arguments
///
/// * `summary` — What was done, in a sentence or two.
///
/// # Errors
///
/// `InvalidArgument` for a blank summary, and `Unavailable` when this session's role ends some other
/// way.
#[doc(alias = "ggop:session.finish")]
pub fn finish(summary: &str) -> Result<(), ApiError> {
    wire::lift(session::finish(summary))
}

/// Accept the work under review: it meets every completion criterion and stays in scope.
///
/// This ends the session. It does not stop the program: whatever follows it still runs.
///
/// # Errors
///
/// `Unavailable` when this session's role ends some other way.
#[doc(alias = "ggop:session.approve")]
pub fn approve() -> Result<(), ApiError> {
    wire::lift(session::approve())
}

/// Reject the work under review, listing every change that must be made before it can be accepted.
///
/// This ends the session, and does not stop the program. Each item says what is wrong and what to
/// change.
///
/// # Arguments
///
/// * `items` — Every change that must be made before the work can be accepted, one per entry: what is
///   wrong, and what to change. It may not be empty.
///
/// # Errors
///
/// `InvalidArgument` when the list is empty, and `Unavailable` when this session's role ends some
/// other way.
#[doc(alias = "ggop:session.request_changes")]
pub fn request_changes(items: &[&str]) -> Result<(), ApiError> {
    wire::lift(session::request_changes(&wire::strings(items)))
}
